src/Lib.P2P/Transports/Tcp.cs

```diff
@@ -0,0 +1,293 @@
+                    var ipaddr = IPAddress.Parse(ip.Value);
+                    await socket.ConnectAsync(ipaddr, port).ConfigureAwait(false);
+                }
+
+                latency = DateTime.Now - start;
+                Log.Trace($"connected to {address} in {latency.TotalMilliseconds} ms");
+            }
+            catch (Exception) when (cancel.IsCancellationRequested)
+            {
+                // eat it, the caller has cancelled and doesn't care.
+            }
+            catch (Exception)
+            {
+                latency = DateTime.Now - start;
+                Log.Trace($"failed to {address} in {latency.TotalMilliseconds} ms");
+                socket?.Dispose();
+                throw;
+            }
+
+            if (cancel.IsCancellationRequested)
+            {
+                Log.Trace("cancel " + address);
+                socket?.Dispose();
+                cancel.ThrowIfCancellationRequested();
+            }
+
+            var timeout = (int) Math.Max(MinReadTimeout.TotalMilliseconds, latency.TotalMilliseconds * 3);
+            socket.LingerState = new LingerOption(false, 0);
+            socket.ReceiveTimeout = timeout;
+            socket.SendTimeout = timeout;
+            Stream stream = new NetworkStream(socket, true);
+            stream.ReadTimeout = timeout;
+            stream.WriteTimeout = timeout;
+
+            stream = new DuplexBufferedStream(stream);
+
+            if (!cancel.IsCancellationRequested)
+            {
+                return stream;
+            }
+            
+            Log.Trace("cancel " + address);
+            await stream.DisposeAsync();
+            cancel.ThrowIfCancellationRequested();
+
+            return stream;
+        }
+
+        /// <inheritdoc />
+        public MultiAddress Listen(MultiAddress address,
+            Action<Stream, MultiAddress, MultiAddress> handler,
+            CancellationToken cancel)
+        {
+            var port = address.Protocols
+               .Where(p => p.Name == "tcp")
+               .Select(p => int.Parse(p.Value))
+               .FirstOrDefault();
+            
+            var ip = address.Protocols
+               .FirstOrDefault(p => p.Name == "ip4" || p.Name == "ip6");
+            
+            if (ip == null)
+            {
+                throw new ArgumentException($"Missing IP address in '{address}'.", nameof(address));
+            }
+            
+            var ipAddress = IPAddress.Parse(ip.Value);
+            var endPoint = new IPEndPoint(ipAddress, port);
+            var socket = new Socket(
+                endPoint.AddressFamily,
+                SocketType.Stream,
+                ProtocolType.Tcp);
+            try
+            {
+                socket.Bind(endPoint);
+                socket.Listen(100);
+            }
+            catch (Exception e)
+            {
+                socket.Dispose();
+                throw new Exception("Bind/listen failed on " + address, e);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: ['System.Exception' should not be thrown by user code.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)